pandas/core/reshape.py

```diff
@@ -232,19 +226,31 @@ def get_new_index(self):
         # construct the new index
         if len(self.new_index_levels) == 1:
             lev, lab = self.new_index_levels[0], result_labels[0]
-            if not (lab == -1).any():
-                return lev.take(lab)
-
-            vals = np.insert(lev.astype('object'), len(lev),
-                             _get_na_value(lev.dtype.type)).take(lab)
-
-            return lev._shallow_copy(vals)
+            return _make_new_index(lev, lab) \
+                    if (lab == -1).any() else lev.take(lab)
 
         return MultiIndex(levels=self.new_index_levels,
                           labels=result_labels,
                           names=self.new_index_names,
                           verify_integrity=False)
 
+
+def _make_new_index(lev, lab):
+    from pandas.core.index import Index, _get_na_value
+
+    nan = _get_na_value(lev.dtype.type)
+    vals = lev.values.astype('object')
+    vals = np.insert(vals, 0, nan) if lab is None else \
+           np.insert(vals, len(vals), nan).take(lab)
+
+    try:
+        vals = vals.astype(lev.dtype, subok=False, copy=False)
+    except ValueError:
+        return Index(vals, **lev._get_attributes_dict())
+
+    return lev._shallow_copy(vals)
+
```

Comment:
``_make_new_index(lev, lab)`` doesn't seem to handle the dtype properly for a ``DatetimeIndex`` when ``lab`` contains ``-1``:

```
In : idx = pd.Index([datetime.now()])
Out:
<class 'pandas.tseries.index.DatetimeIndex'>
[2015-02-12 16:00:12.487944]
Length: 1, Freq: None, Timezone: None

In : _make_new_index(idx, None)
Out: Index([NaT, 1423756812487944000], dtype='object')

In : _make_new_index(idx, [0])
Out:
[2015-02-12 16:00:12.487944]
Length: 1, Freq: None, Timezone: None

In : _make_new_index(idx, [0, -1])
Out: Index([1423756812487944000, NaT], dtype='object')
```
I need this in order to get ``stack()`` to work properly with ``NaN``/``NaT`` in indices for https://github.com/pydata/pandas/pull/9023.